client/src/lib.rs

```diff
@@ -134,8 +141,94 @@ pub struct ClientOptions {
     pub keep_alive: Option<ClientKeepAliveConfig>,
 }
 
+/// Connection options that can be dynamically updated. Construct with [ClientOptionsUpdateBuilder].
+/// All fields are optional, and will be ignored if not set. To disable optional fields in [ClientOptions],
+/// set field to a tombstone specified in the field description.
+#[derive(Clone, Debug, Default, derive_builder::Builder)]
+#[non_exhaustive]
+pub struct ClientOptionsUpdate {
```

Comment:
Can the entire `ClientOptions` struct be reused instead? Not every field has to be optional. You can require the callers provide the same data they did for original client options if they don't want to change it. The lang side doesn't need help here.

(would also be neat if some of the runtime options like headers was set/updated via a common mechanism and didn't require reconnect, but that doesn't have to be now)

Reply:
I think this probably makes decent sense too. I imagine the lang-side API will typically involve the user copying some existing options and changing them, in which case this all works out fine.

Reply:
These are the only fields that I can change dynamically, because they are associated with the endpoint, and I think it makes it more clear what can be changed with a separate struct. Also, fields like `target_url` are not optional in ClientOptions but need to be optional in update, or remember the original value set, which is a pain with concurrent updates...

Reply:
If we go with swapping out the whole options behind a mutex which I mentioned in another comment, then the other fields should be changeable too I think, which could work out nicely.

Reply:
Things like `retry_config` are used by the Retry wrapper that I cannot change.  `client_name` or `client_version` or `identity` may be used in metrics, we probably don't want to change them, and I cannot see a use case to change them dynamically... We could fail the update if they don't match. 
I still think that is easier in the API to do deltas with only the changeable attributes, but internally we could use ClientOptions instead of a changeset, if we are ok with a mutex on options. We would still need some state for the updates that does not belong in options, such as the channel, the version id,... but we could probably lock them independently from options...

Reply:
Why can't they be changed? I agree if they can't be changed it makes sense to accept something that only allows changing what can be. But, I'm not sure I follow why we couldn't change those. They are accessed each time a request is made - seems like changing them would affect subsequent requests and work properly?

Reply:
for retry_config we cache it inside:
impl<SG> RetryClient<SG> {
    /// Use the provided retry config with the provided client
    pub fn new(client: SG, retry_config: RetryConfig) -> Self {
        Self {
            client,
            retry_config: Arc::new(retry_config),
        }
    }
}
we could change that if needed, but not sure if it is that important to dynamically change retry_config...

The others we can, but the question is whether we should, if it is used in metrics, or server side, it may be confusing, and
I don't see much value of changing them dynamically...

Reply:
We don't have to actually expose those to end users, but I take your point. Especially since the `RetryClient` is a level higher. I'm fine with having just the diff struct.